src/db/database.ts

```diff
@@ -155,27 +164,50 @@ export class Database<T extends Schema> {
         );
         const iter = btree.iter(valueRef);
 
+        // for a given document, the number of times this specific token appears
+        let tfMap = new Map<DataPointer, number>();
         while (await iter.next()) {
           const currentKey = iter.getKey();
 
           if (ReferencedValue.compareBytes(currentKey.value, valueBuf) !== 0) {
             break;
           }
           const mp = iter.getPointer();
-          const data = await this.dataFile.get(
-            Number(mp.offset),
-            Number(mp.offset) + mp.length - 1,
-          );
-          table.insert(data);
+
+          const dp: DataPointer = {
+            start: Number(mp.offset),
+            end: Number(mp.offset) + mp.length - 1,
+          };
+
+          tfMap.set(dp, (tfMap.get(dp) ?? 0) + 1);
         }
-      }
 
-      const data = table.top;
+        // BM25 Algorithm
+        // https://docs.vespa.ai/en/reference/bm25.html#:~:text=The%20bm25%20rank%20feature%20implements,over%20an%20indexed%20string%20field.
+        const n = entries; // total number of documents
+        const nt = tfMap.size; // number of documents containing this token
+        const idf = Math.log((n - nt + 0.5) / (nt + 0.5)); // inverse document frequency formula
+
+        // BM25 constants
+        const K1 = 1.2;
+        const B = 0.75;
+        const avgdl = totalFieldValueLength / entries; // Average document length
+
+        for (const [key, tf] of tfMap.entries()) {
+          const { start, end } = key;
+          const data = await this.dataFile.get(start, end);
+          const dataRecord = JSON.parse(data);
+          const valueLength = dataRecord[fieldName].length;
 
-      if (!data) {
-        throw new Error(`no trigrams were evaluated`);
+          const num = tf * (K1 + 1);
+          const den = tf + K1 * (1 - B + (B * valueLength) / avgdl);
+          const score = idf * (num / den);
+
+          table.insert(data, score);
+        }
       }
-      yield handleSelect(data.key, query.select);
+
+      yield table.top();
```

Comment:
```suggestion
      yield* table.top();
```

yield individual results